components/brave_wallet/common/zcash_utils.h

```diff
@@ -89,21 +104,218 @@ struct OrchardOutput {
+  // List of note positions that were spent at this checkpoint.
+  std::vector<uint32_t> marks_removed;
+};
+
+struct OrchardCheckpointBundle {
+  OrchardCheckpointBundle(uint32_t checkpoint_id, OrchardCheckpoint);
+  ~OrchardCheckpointBundle();
+  OrchardCheckpointBundle(const OrchardCheckpointBundle& other);
+  OrchardCheckpointBundle& operator=(const OrchardCheckpointBundle& other);
+  OrchardCheckpointBundle(OrchardCheckpointBundle&& other);
+  OrchardCheckpointBundle& operator=(OrchardCheckpointBundle&& other);
+
+  bool operator==(const OrchardCheckpointBundle& other) const = default;
+
+  // The block height serves as the checkpoint identifier.
+  uint32_t checkpoint_id;
+  OrchardCheckpoint checkpoint;
+};
+
+// Address of a subtree in the shard tree.
+struct OrchardShardAddress {
+  uint8_t level = 0;
+  uint32_t index = 0;
+
+  bool operator==(const OrchardShardAddress& other) const = default;
+};
+
+// Top part of the shard tree from the root to the shard roots level
+// Used for optimization purposes in the shard tree crate.
+struct OrchardCap {
+  OrchardCap();
+  ~OrchardCap();
+
+  OrchardCap(const OrchardCap& other);
+  OrchardCap& operator=(const OrchardCap& other);
+  OrchardCap(OrchardCap&& other);
+  OrchardCap& operator=(OrchardCap&& other);
+
+  std::vector<uint8_t> data;
+};
+
+// Subtree with root selected from the shard roots level.
+struct OrchardShard {
+  OrchardShard();
+  OrchardShard(OrchardShardAddress shard_addr,
+               std::optional<OrchardShardRootHash> shard_hash,
+               std::vector<uint8_t> shard_data);
+  ~OrchardShard();
+
+  OrchardShard(const OrchardShard& other);
+  OrchardShard& operator=(const OrchardShard& other);
+  OrchardShard(OrchardShard&& other);
+  OrchardShard& operator=(OrchardShard&& other);
+
+  bool operator==(const OrchardShard& other) const = default;
+
+  // Subtree root address.
+  OrchardShardAddress address;
+  // Root hash exists only on completed shards.
+  std::optional<OrchardShardRootHash> root_hash;
+  std::vector<uint8_t> shard_data;
+  // Right-most position of the subtree leaf.
+  size_t subtree_end_height = 0;
+};
+
+struct OrchardCommitment {
+  OrchardCommitmentValue cmu;
+  bool is_marked;
+  std::optional<uint32_t> checkpoint_id;
+};
+
+// Compact representation of the Merkle tree on some point.
+// Since batch inserting may contain gaps between scan ranges we insert
+// frontier which allows to calculate node hashes and witnesses(merkle path from
+// leaf to the tree root) even when previous scan ranges are not completed.
+struct OrchardTreeState {
+  OrchardTreeState();
+  ~OrchardTreeState();
+  OrchardTreeState(const OrchardTreeState&);
+
+  // Tree state is linked to the end of some block.
+  uint32_t block_height = 0u;
+  // Number of leafs at the position.
+  uint32_t tree_size = 0u;
+  // https://docs.aztec.network/protocol-specs/l1-smart-contracts/frontier
+  std::vector<uint8_t> frontier;
+};
+
+class OrchardShardTreeDelegate {
```

Comment:
This definitely should go in its own file

Reply:
Moved to separate file